### PR TITLE
MRG, BUG: Use auto extrap and mean borders

### LINKED ISSUE #7601: BUG: Extreme topomap valuse still too strong
@agramfort pointed out that the `examples/time_frequency/plot_source_power_spectrum_opm.py` example had edges that are too strong:

master

Doing `border='mean'` plus `extrapolate='local'` gives:

Screenshot from 2020-04-13 09-44-01

I wonder if:

1. The default extrapolation mode of `box` should actually add the local points, then also add the four points in the corners. It might be a good compromise.
2. `border='mean'` should be the default

If we did these two things, this is what the new default would look like (hacked together the box change for now and set `border='mean'`):

Screenshot from 2020-04-13 10-03-41

I know we've had some conversations about (2) at least, but I'd also like (1) because it might combine the local focality of `'local'` with the smoothness (bounded by matplotlib mask instead of interpolation grid extent) of `'box'`. Thoughts @mmagnuski ?

### PR DESCRIPTION
**Pull request by larsoner**
Probably a better solution than #7602.

WIP because:

- [x] viz examples/tutorials need to be updated
- [ ] [examples](https://19301-1301584-gh.circle-artifacts.com/0/dev/index.html) need to be checked

Closes #7601
Closes #7602

### CODE REVIEW

**Comment by larsoner** [comment]
[Examples have rendered](https://19301-1301584-gh.circle-artifacts.com/0/dev/index.html) in case anyone wants to look

**Comment by larsoner** [comment]
Okay the MEG looks reasonable, but @mmagnuski you are right that the local convex hull is not always good -- for EEG in `plot_evoked_topomap` it does not look good.

This makes me wonder if we want to add `extrapolate='auto'`, which means `'local'` for MEG and `'head'` for EEG. Or maybe we want to add some softening / expanding of the edges, but that's probably going to be tricky to get right.

**Comment by drammock** [comment]
The renderings in [this example](https://19301-1301584-gh.circle-artifacts.com/0/dev/auto_examples/visualization/plot_evoked_topomap.html) look good to me. The renderings in [this example](https://19301-1301584-gh.circle-artifacts.com/0/dev/auto_examples/io/plot_read_proj.html) also look good as far as the "local"/convex hull question goes, but it looks like the head outline is slightly off-center left-to-right?  I realize that's not what you're addressing in this PR, but it's rather puzzling since both examples use the sample data.

**Comment by agramfort** [comment]
MEG topos look much nicer:

https://19301-1301584-gh.circle-artifacts.com/0/dev/auto_examples/io/plot_read_proj.html#sphx-glr-auto-examples-io-plot-read-proj-py

but our EEG friends may have some concerns.

extrapolate='auto' is maybe a good option.

@mmagnuski @cbrnr thoughts @sappelhoff thoughts?

**Comment by agramfort** [comment]
it's still a bit unclear to me why the EEG sensors on sample are so
centered inside the head circle. There
are electrodes closer to the ears in 3d.

>

**Comment by larsoner** [comment]
They are that way in `master`, it's just harder to see when the default interp goes to the circle

- [master](https://mne.tools/dev/auto_examples/visualization/plot_evoked_topomap.html#additional-plot-topomap-options)
- [PR](https://19309-1301584-gh.circle-artifacts.com/0/dev/auto_examples/visualization/plot_evoked_topomap.html#additional-plot-topomap-options) (unchanged pos)

It's possible this will be fixed by #7455, but I'd rather deal with the problem separately since it's orthogonal and not added by this PR.

**Comment by mmagnuski** [comment]
@larsoner
> This makes me wonder if we want to add extrapolate='auto', which means 'local' for MEG and 'head' for EEG. Or maybe we want to add some softening / expanding of the edges, but that's probably going to be tricky to get right.

I  don't think an auto of `extrapolate='head'` would be good - as this generates artifacts when some channels are outside of the head circle. As such cases are pretty common now - this might lead to more surprise than the proposed always 'local' scenario. (`extrapolate='circle'` might be a good alternative to what 'head' does currently, but that's another issue)
When it comes to softening etc. - that's why I was referring to using two mask at the same time: the circle and convex hull. The circle smoothes and the convex hull removes whatever is not masked by the circle. But it would likely require some tweaking, and I know you don't like this idea very much, so I'm not pressing. :)

There seems to be some pixelization not masked that is visible here (third topo from the left):
image
but it seems related to the low `res` in that code cell (`res=32`).

**Comment by larsoner** [comment]
> When it comes to softening etc. - that's why I was referring to using two mask at the same time: the circle and convex hull. The circle smoothes and the convex hull removes whatever is not masked by the circle. But it would likely require some tweaking, and I know you don't like this idea very much, so I'm not pressing. :)

This will not work for the case linked above -- you have a (correct) hexagonal convex hull, so adding an additional circular mask on top of this will either be a no-op or it will remove some sensors. If we want to keep `local` the default but soften it, a better approach might be to "push" some points out a bit from the radius. Or figure out how to get Bezier points or something... but I'd rather avoid these complications if possible.

> There seems to be some pixelization not masked that is visible here (third topo from the left):

Ahh yes, we'll need to tweak how far out we put the extra points, how many of them we have, etc. I changed some values here that might have made the "jagginess" problem worse, but we'll have to evaluate that in the context of not introducing "blobs", too.

Ultimately if someone uses a low res, though, I don't mind them ending up with jagged lines...

> I don't think an auto of extrapolate='head' would be good - as this generates artifacts when some channels are outside of the head circle. 

Where do you see this? My understanding is that "head" really means "minimum of the head circle and the circle that encompasses all sensors", so it should always give at least a somewhat sensible one.

It seems like the simplest solution is to make the default `head` for EEG and `local` for MEG, and then fix the problems with `head` plotting for EEG (since we need to do those anyway.

**Comment by mmagnuski** [comment]
> This will not work for the case linked above -- you have a (correct) hexagonal convex hull, so adding an additional circular mask on top of this will either be a no-op or it will remove some sensors. If we want to keep local the default but soften it, a better approach might be to "push" some points out a bit from the radius. Or figure out how to get Bezier points or something... but I'd rather avoid these complications if possible.

I didn't mean this in the context of the example I pasted. If you increase the distance between channels and extrapolation points (and thus the distance of convex hull mask) the circle mask trick would work. But I understand your point, and it is a small thing - its good as it is.

> Where do you see this? My understanding is that "head" really means "minimum of the head circle and the circle that encompasses all sensors", so it should always give at least a somewhat sensible one.

It is more visible with `border=0`, but you can see that the extrapolation points are at the head circle:
image
with `border='mean'` it's less visible but the extrapolation points are still at the head circle. I think they should be where the circle mask is (or slightly outside to be precise) - but some time ago we discussed this and you preferred it stayed at the head circle.

> It seems like the simplest solution is to make the default head for EEG and local for MEG, and then fix the problems with head plotting for EEG (since we need to do those anyway.

Sure, if we fix `extrapolate='head'` then it would be ok. :)

> Ultimately if someone uses a low res, though, I don't mind them ending up with jagged lines...

:+1:

**Comment by cbrnr** [comment]
Plotting IC topomaps with my limited set of channels (I still think they should be closer to the head circle but this is the other issue we've not solved) doesn't look great with the new defaults:

Screen Shot 2020-04-14 at 12 03 50

I'd prefer if we'd extrapolate at least to the head circle.

**Comment by hoechenberger** [comment]
@cbrnr 
> I'd prefer if we'd extrapolate at least to the head circle.

+1 on that one

**Comment by agramfort** [comment]
what do we do here to move on?

**Comment by larsoner** [comment]
My vote is to do `extrapolate='auto'` which means 'local' for MEG and 'head' for EEG. Also maybe try to reduce the "edge blobs" here if reasonable to do so. Then I think the remaining issues can be dealt with in separate PRs (probably mostly for EEG / head / skirt modes?).

**Comment by agramfort** [comment]
that works for me !

>

**Comment by larsoner** [comment]
> I think they should be where the circle mask is (or slightly outside to be precise) - but some time ago we discussed this and you preferred it stayed at the head circle.

> Sure, if we fix extrapolate='head' then it would be ok. :)

Okay done now along with `'auto'` mode. Running `[circle full]` we'll see how it looks!

**Comment by larsoner** [comment]
I have not looked at all examples, but this looks okay at least. If others agree I can take a look at all examples:

https://19486-1301584-gh.circle-artifacts.com/0/dev/auto_examples/visualization/plot_evoked_topomap.html#additional-plot-topomap-options

**Comment by mmagnuski** [comment]
Looks ok. So, in a subsequent PR we can change the behavior of extrapolate='head' so that extrapolation points are not placed on the head outline if some channels extend beyond this outline, right?

**Comment by larsoner** [comment]
It already should extend outside the head, which is what you wanted, right?

**Comment by mmagnuski** [comment]
Ah, you changed it? Sorry, didn't notice.

File: mne/viz/topomap.py [hunk 1]
```diff
@@ -546,18 +559,19 @@ def _get_extra_points(pos, extrapolate, sphere):
...
     else:
+        assert extrapolate == 'head'
         # return points on the head circle
-        angle = np.arcsin(distance / 2 / head_radius)
+        angle = np.arcsin(distance / 2 / np.mean(radii))
         points_l = np.arange(0, 2 * np.pi, angle)
-        use_radius = head_radius * 1.1
-        points_x = np.cos(points_l) * use_radius + x
-        points_y = np.sin(points_l) * use_radius + y
+        use_radii = radii * 1.1
```
Comment (mmagnuski) [inline mne/viz/topomap.py:566]: Ok, so `radii` are x and y clip radius now, not the head_radius, 👍

Comment (larsoner) [reply mne/viz/topomap.py:566]: Yes indeed!

**Comment by cbrnr** [comment]
My ICA components example looks good!

**Comment by sappelhoff** [comment]
Had a quick look at some examples, looks good to me as well :+1: 

Could you post your example once more with the new settings @cbrnr ?

**Comment by cbrnr** [comment]
Here are my IC topomaps - except for the issue that these now look much more compressed than before (#7472 and #7141) I'm happy:

Screen Shot 2020-04-21 at 09 38 11

**Comment by mmagnuski** [comment]
@larsoner
Could we just change the docs for `extrapolate='head'` so that it is clear that the extrapolation is to the head outline or outside?

**Comment by hoechenberger** [comment]
Looks great, awesome work!

**Comment by larsoner** [comment]
@mmagnuski changed in https://github.com/mne-tools/mne-python/pull/7603/commits/15e5baef6c447e2d1cb3864cae09178ffb12a3ca#diff-c0cc0663d36fcb3b69d6e33a22573db1R87-R90 to:

        Extrapolate out to the edges of the clipping circle. This will be on
        the head circle when the sensors are contained within the head circle,
        but it can extend beyond the head when sensors are plotted outside
        the head circle.

Okay?

**Comment by mmagnuski** [comment]
Perfect!

**Comment by larsoner** [comment]
Okay, hopefully it's easier to fix the remaining issues now!

**Comment by sappelhoff** [comment]
Thanks @larsoner!

**Pull request merged**